File: Daily/sbgo/Leetcode915-公交路线.go
package sbgo

func numBusesToDestination(routes [][]int, source int, target int) int {
	if target == source {
		return 0
	}
	n := len(routes)
	dist := make([]int, n)
	for i := 0; i < n; i++ {
		dist[i] = int(1e8)
	}
	que := []int{}
	g := map[int][]int{}
	for i := 0; i < n; i++ {
		for _, j := range routes[i] {
			if j == source {
				que = append(que, i)
				dist[i] = 1
			}
			g[j] = append(g[j], i)
		}
	}

	for len(que) > 0 {
		t := que[0]
		que = que[1:]
		for _, x := range routes[t] {
			if x == target {
				return dist[t]
			}
			for _, y := range g[x] {
				if dist[y] > dist[t]+1 {
					dist[y] = dist[t] + 1
					que = append(que, y)
				}
			}
			delete(g, x)
		}
	}
	return -1
}

File: Daily/Rust/Leetcode915-公交路线.rs
use std::collections::HashMap;
impl Solution {
    pub fn num_buses_to_destination(routes: Vec<Vec<i32>>, source: i32, target: i32) -> i32 {
        if target == source{
            return 0;
        }
        let n = routes.len();
        let mut que = Vec::new();
        let mut g = HashMap::new();
        let mut dist = vec![1e8 as i32;n];
        for i in 0..n{
            for &j in routes[i].iter(){
                if j == source{
                    que.push(i);
                    dist[i] = 1;
                }
                g.entry(j).or_insert(vec![]).push(i as i32);
            }
        }

        while que.len() > 0{
            let t = que[0];
            que.remove(0);
            for &x in routes[t as usize].iter(){
                if x == target{
                    return dist[t as usize]; 
                }
                if g.contains_key(&x){
                    for &y in g[&x].iter(){
                        if dist[y as usize] > dist[t as usize] + 1{
                            dist[y as usize] = dist[t as usize] + 1;
                            que.push(y as usize);
                        }
                    }
                    g.remove_entry(&x);
                }
            }
        }
        -1
    }
}